File: src/op/linear.cc
#include "op/linear.h"

#include <cassert>
#include <tuple>

#include "tensor/tensor.cuh"
#include "backend/cuda/linear.cuh"

namespace TinyTorch {

Tensor linear_forward(const Tensor &input, const Tensor &weight, const Tensor &bias) {
    assert(input.device.is_cuda() && weight.device.is_cuda() && bias.device.is_cuda());
    assert(input.dim() == 2 &&
           weight.dim() == 2 &&
           bias.dim() == 1);
    assert(input.shape[1] == weight.shape[0] &&
           weight.shape[1] == bias.shape[0]);
    int batchsize = input.shape[0];
    int in_features = weight.shape[0];
    int out_features = weight.shape[1];
    Tensor output({batchsize, out_features}, input.device);
    if (input.device.is_cuda()) {
        Backend::CUDA::linear_forward(input.data, output.data, weight.data, bias.data,
                                      batchsize, in_features, out_features);
    } else {}
    return output;
}

std::tuple<Tensor, Tensor, Tensor> linear_backward(
        const Tensor &input,
        const Tensor &weight,
        const Tensor &grad_output) {
    assert(input.device.is_cuda() &&
           weight.device.is_cuda() &&
           grad_output.device.is_cuda());
    assert(input.dim() == 2 &&
           weight.dim() == 2 &&
           grad_output.dim() == 2);
    assert(input.shape[0] == grad_output.shape[0] &&
           input.shape[1] == weight.shape[0] &&
           grad_output.shape[1] == weight.shape[1]);
    int batchsize = input.shape[0];
    int in_features = weight.shape[0];
    int out_features = weight.shape[1];
    Tensor grad_input(input.shape, input.device);
    Tensor grad_weight(weight.shape, input.device);
    Tensor grad_bias({out_features}, input.device);
    if (input.device.is_cuda()) {
        Backend::CUDA::linear_backward(input.data, nullptr, weight.data, nullptr,
                                       batchsize, in_features, out_features,
                                       grad_output.data, grad_input.data, grad_weight.data, grad_bias.data);
    } else {}
    return std::make_tuple(std::move(grad_input), std::move(grad_weight), std::move(grad_bias));
}

}

File: unittest/linear.py
import tinytorch

import torch
from torch import nn
from torch.nn import functional as F

import numpy as np

BATCHSIZE = 10
IN_FEATURES = 32
OUT_FEATURES = 64

device = None

def test_linear():
    input = torch.randn((BATCHSIZE, IN_FEATURES), device=device, requires_grad=True)
    weight = torch.randn((OUT_FEATURES, IN_FEATURES), device=device, requires_grad=True)
    bias = torch.randn(OUT_FEATURES, device=device, requires_grad=True)
    grad_output = torch.randn((BATCHSIZE, OUT_FEATURES), device=device, requires_grad=True)
    output_gt = F.linear(input, weight, bias)
    output_gt.backward(grad_output)

    output_gt = tinytorch.TensorBase(output_gt.detach().numpy())
    grad_input_gt = tinytorch.TensorBase(input.grad.detach().numpy())
    grad_weight_gt = tinytorch.TensorBase(weight.grad.transpose(0, 1).detach().numpy())
    grad_bias_gt = tinytorch.TensorBase(bias.grad.detach().numpy())
    input = tinytorch.TensorBase(input.detach().numpy())
    weight = tinytorch.TensorBase(weight.transpose(0, 1).detach().numpy())
    bias = tinytorch.TensorBase(bias.detach().numpy())
    grad_output = tinytorch.TensorBase(grad_output.detach().numpy())

    output_res = tinytorch.op.linear_forward(input, weight, bias)
    grad_input_res, grad_weight_res, grad_bias_res = tinytorch.op.linear_backward(input, weight, grad_output)

    assert output_res == output_gt
    assert grad_input_res == grad_input_gt
    assert grad_weight_res == grad_weight_gt
    assert grad_bias_res == grad_bias_gt

def main():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print("Torch Device:", device)
    tinytorch.Device.set_default_device(tinytorch.Device.cuda())
    print("TinyTorch Default Device:", "cpu" if tinytorch.Device.get_default_device().is_cpu() else "cuda")
    test_linear()

if __name__ == "__main__":
    main()
